packages/kit/src/runtime/client/client.js

```diff
@@ -1902,12 +1901,40 @@ function reset_focus() {
 			root.removeAttribute('tabindex');
 		}
 
-		return new Promise((resolve) => {
+		// capture current selection, so we can compare the state after
+		// snapshot restoration and afterNavigate callbacks have run
+		const selection = getSelection();
+
+		if (selection && selection.type !== 'None') {
+			/** @type {Range[]} */
+			const ranges = [];
+
+			for (let i = 0; i < selection.rangeCount; i += 1) {
+				ranges.push(selection.getRangeAt(i));
+			}
+
 			setTimeout(() => {
+				if (selection.rangeCount !== ranges.length) return;
+
+				for (let i = 0; i < selection.rangeCount; i += 1) {
+					const a = ranges[i];
+					const b = selection.getRangeAt(i);
+
+					// we need to do a deep comparison rather than just `a !== b` because
+					// Safari behaves differently to other browsers
+					if (a.commonAncestorContainer !== b.commonAncestorContainer) return;
+					if (a.startContainer !== b.startContainer) return;
+					if (a.endContainer !== b.endContainer) return;
+					if (a.startOffset !== b.startOffset) return;
+					if (a.endOffset !== b.endOffset) return;
```

Comment:
I think we can code golf a bit at least by turning this into one big `if(.. !== .. || ..) return`